contracts/src/token/erc1155/extensions/supply.rs

```diff
@@ -0,0 +1,521 @@
+//! Extension of ERC-1155 that adds tracking of total supply per token id.
+//!
+//! Useful for scenarios where Fungible and Non-fungible tokens have to be
+//! clearly identified. Note: While a `_total_supply` of 1 might mean the
+//! corresponding is an NFT, there are no guarantees that no other tokens
+//! with the same id are not going to be minted.
+//!
+//! NOTE: This contract implies a global limit of 2**256 - 1 to the number
+//! of tokens that can be minted.
+//!
+//! CAUTION: This extension should not be added in an upgrade to an already
+//! deployed contract.
+
+use alloc::{vec, vec::Vec};
+
+use alloy_primitives::{Address, U256};
+use openzeppelin_stylus_proc::interface_id;
+use stylus_sdk::{
+    abi::Bytes,
+    msg,
+    prelude::{public, sol_storage},
+};
+
+use crate::{
+    token::erc1155::{self, Erc1155, IErc1155},
+    utils::math::storage::SubAssignUnchecked,
+};
+
+sol_storage! {
+    /// State of an [`Erc1155Supply`] token.
+    pub struct Erc1155Supply {
+        /// ERC-1155 contract storage.
+        Erc1155 erc1155;
+        /// Mapping from token id to total supply.
+        mapping(uint256 => uint256) _total_supply;
+        /// Total supply of all token ids.
+        uint256 _total_supply_all;
+    }
+}
+
+/// Required interface of a [`Erc1155Supply`] contract.
+#[interface_id]
+pub trait IErc1155Supply {
+    /// Total value of tokens in with a given id.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the contract's state.
+    /// * `id` - Token id as a number.
+    fn total_supply(&self, id: U256) -> U256;
+
+    /// Total value of tokens.
+    ///
+    /// # Arguments
+    ///
+    /// * `&self` - Read access to the contract's state.
+    #[selector(name = "totalSupply")]
+    fn total_supply_all(&self) -> U256;
```

Comment:
You can indicate in rustdocs: once user reexports the function, it should have selector renamed as "totalSupply".